src/CachingDocBlockFactory.php

```diff
@@ -0,0 +1,72 @@
+<?php
+declare(strict_types=1);
+
+namespace LanguageServer;
+
+use Microsoft\PhpParser\Node;
+use phpDocumentor\Reflection\DocBlock;
+use phpDocumentor\Reflection\DocBlockFactory;
+use phpDocumentor\Reflection\Types;
+
+/**
+ * Caches DocBlocks by node start position and file URI.
+ */
+class CachingDocBlockFactory
+{
+    /**
+     * Maps file + node start positions to DocBlocks.
```

Comment:
In the pre-tolerant parser version I used to save the docblock instance on the Node itself so they would get naturally get cached and garbage collected when the document gets garbage collected. The `CachingDocBlockFactory` instance is saved in the DefinitionResolver, which is a long-living singleton-like object. Why not save this map on `PhpDocument` instances? That way it is not a cache that needs to be cleared on seemingly unrelated events like new messages, it is simply a memoization store to remember lazily computed docblocks. It might actually be just as fast to eagerly compute them all because we need to do that anyway to get the hover contents for the index.

Reply:
This makes a lot of sense. Perhaps the name resolution cache from the Scope thing could also live in `PhpDocument` then?